sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala

```diff
@@ -57,20 +60,22 @@ private[sql] class RocksDBStateStoreProvider
         keyStateEncoderSpec: KeyStateEncoderSpec,
         useMultipleValuesPerKey: Boolean = false,
         isInternal: Boolean = false): Unit = {
-      verify(colFamilyName != StateStore.DEFAULT_COL_FAMILY_NAME,
-        s"Failed to create column family with reserved_name=$colFamilyName")
-      verify(useColumnFamilies, "Column families are not supported in this store")
-      rocksDB.createColFamilyIfAbsent(colFamilyName, isInternal)
+      ColumnFamilyUtils.createColFamilyIfAbsent(colFamilyName, isInternal)
+
       keyValueEncoderMap.putIfAbsent(colFamilyName,
-        (RocksDBStateEncoder.getKeyEncoder(keyStateEncoderSpec),
+        (RocksDBStateEncoder.getKeyEncoder(keyStateEncoderSpec, useColumnFamilies),
```

Comment:
Wait, we have encoder per column family. Do we have a case where we assign different IDs to the same CF during lifecycle of the encoder instance? 

If we don't, we can require encoder to maintain the CF ID it is in charge of, and caller no longer needs to look up twice for finding encoder vs finding CF ID. This can be done along with the above proposal - abstract class can handle this. If this is done, then we can remove param for vcfId and base trait to be unchanged.

Even if we do, if we can re-initialize the encoder for new CF ID, then the above can be still applied.

Reply:
We won't have case where we assign different IDs to the same CF during lifecycle of the encoder instance. In my new refactored RocksDBKeyEncoder class, we'll pass in a virtual column family id during column family creation time. This way we can get rid of looking up column family id every time in the `RocksDBStateStoreProvider`.